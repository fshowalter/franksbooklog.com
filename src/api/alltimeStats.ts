import { perfLogger } from "~/utils/performanceLogger";

import type { AlltimeStatsJson } from "./data/alltimeStatsJson";

import { alltimeStatsJson } from "./data/alltimeStatsJson";

export type AlltimeStats = AlltimeStatsJson & {};

export async function alltimeStats(): Promise<AlltimeStats> {
  return await perfLogger.measure("alltimeStats", async () => {
    return await alltimeStatsJson();
  });
}
