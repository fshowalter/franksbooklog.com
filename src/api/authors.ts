import { allAuthorsJson, type AuthorJson } from "./data/authorsJson";

export type Author = AuthorJson & {};

type AuthorDetails = {
  author: Author;
  distinctKinds: string[];
  distinctPublishedYears: string[];
};

export async function allAuthors(): Promise<Author[]> {
  return await allAuthorsJson();
}

export async function getAuthorDetails(slug: string): Promise<AuthorDetails> {
  const authors = await allAuthorsJson();
  const distinctKinds = new Set<string>();
  const distinctPublishedYears = new Set<string>();

  const author = authors.find((value) => value.slug === slug)!;

  for (const work of author.works) {
    distinctKinds.add(work.kind);
    distinctPublishedYears.add(work.yearPublished);
  }

  return {
    author,
    distinctKinds: [...distinctKinds].toSorted(),
    distinctPublishedYears: [...distinctPublishedYears].toSorted(),
  };
}
