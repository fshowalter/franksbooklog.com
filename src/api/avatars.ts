import { getImage } from "astro:assets";

import { normalizeSources } from "./utils/normalizeSources";

export interface AvatarImageProps {
  src: string;
  srcSet: string;
}

const images = import.meta.glob<{ default: ImageMetadata }>(
  "/content/assets/avatars/*.png",
);

async function getAvatarFile(slug: string) {
  const imagePath = Object.keys(images).find((path) => {
    return path.endsWith(`${slug}.png`);
  });

  if (!imagePath) {
    return null;
  }

  return await images[imagePath]();
}

export async function getAvatarImageProps(
  slug: string,
  { width, height }: { width: number; height: number },
): Promise<AvatarImageProps | null> {
  const avatarFile = await getAvatarFile(slug);

  if (!avatarFile) {
    return null;
  }

  const optimizedImage = await getImage({
    src: avatarFile.default,
    width: width,
    height: height,
    format: "avif",
    densities: [1, 2],
    quality: 80,
  });

  return {
    srcSet: normalizeSources(optimizedImage.srcSet.attribute),
    src: normalizeSources(optimizedImage.src),
  };
}
