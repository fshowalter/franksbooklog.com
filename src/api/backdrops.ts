import { getImage } from "astro:assets";

import { normalizeSources } from "./utils/normalizeSources";

export type BackdropImageProps = {
  src: string;
  srcSet: string;
};

const images = import.meta.glob<{ default: ImageMetadata }>(
  "/content/assets/backdrops/*.png",
);

export async function getBackdropImageProps(
  slug: string,
  {
    height,
    width,
  }: {
    height: number;
    width: number;
  },
): Promise<BackdropImageProps> {
  const backdropFile = await getBackdropFile(slug);

  const optimizedImage = await getImage({
    format: "avif",
    height: height,
    quality: 80,
    src: backdropFile.default,
    width: width,
    widths: [0.3, 0.5, 0.8, 1].map((w) => w * width),
  });

  return {
    src: normalizeSources(optimizedImage.src),
    srcSet: normalizeSources(optimizedImage.srcSet.attribute),
  };
}

async function getBackdropFile(slug: string) {
  let backdropFilePath = Object.keys(images).find((path) => {
    return path.endsWith(`${slug}.png`);
  })!;

  backdropFilePath =
    backdropFilePath ||
    Object.keys(images).find((path) => {
      return path.endsWith(`default.png`);
    })!;

  return await images[backdropFilePath]();
}
