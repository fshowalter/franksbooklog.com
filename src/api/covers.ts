import { getImage } from "astro:assets";
import fs from "node:fs";
import path from "node:path";
import sharp from "sharp";

import {
  createCacheConfig,
  createCacheKey,
  ensureCacheDir,
  getCachedItem,
  saveCachedItem,
} from "~/utils/cache";

import { normalizeSources } from "./utils/normalizeSources";

export type CoverImageProps = {
  height: number;
  src: string;
  srcSet: string;
  width: number;
};

type Work = {
  slug: string;
};

const images = import.meta.glob<{ default: ImageMetadata }>(
  "/content/assets/covers/*.png",
);

const cacheConfig = createCacheConfig("cover-base64");

export async function getCoverHeight(coverPath: string, targetWidth: number) {
  try {
    const { height, width } = await sharp(coverPath).metadata();

    return (height / width) * targetWidth;
  } catch (error) {
    console.error("Error:", error);
    return 0;
  }
}

export async function getCoverWidth(work: Work, targetHeight: number) {
  try {
    const coverPath = getWorkCoverPath(work);
    const { height, width } = await sharp(coverPath).metadata();

    return (width / height) * targetHeight;
  } catch (error) {
    console.error("Error:", error);
    return 0;
  }
}

export async function getFeedCoverProps(work: Work): Promise<CoverImageProps> {
  const workCoverFile = await getWorkCoverFile(work);

  const optimizedImage = await getImage({
    format: "jpeg",
    height: 750,
    quality: 80,
    src: workCoverFile.default,
    width: 500,
  });

  return {
    height: 750,
    src: normalizeSources(optimizedImage.src),
    srcSet: normalizeSources(optimizedImage.srcSet.attribute),
    width: 500,
  };
}

export async function getFixedCoverImageProps(
  work: Work,
  { width }: { width: number },
): Promise<CoverImageProps> {
  const workCoverFile = await getWorkCoverFile(work);
  const workCoverPath = getWorkCoverPath(work);

  const height = await getCoverHeight(workCoverPath, width);

  const optimizedImage = await getImage({
    densities: [1, 2],
    format: "avif",
    height,
    quality: 80,
    src: workCoverFile.default,
    width: width,
  });

  return {
    height,
    src: normalizeSources(optimizedImage.src),
    srcSet: normalizeSources(optimizedImage.srcSet.attribute),
    width: width,
  };
}

export async function getFluidCoverImageProps(
  work: Work,
  { width }: { width: number },
): Promise<CoverImageProps> {
  const workCoverFile = await getWorkCoverFile(work);
  const coverFilePath = getWorkCoverPath(work);
  const height = await getCoverHeight(coverFilePath, width);

  const optimizedImage = await getImage({
    format: "avif",
    height,
    quality: 80,
    src: workCoverFile.default,
    width: width,
    widths: [0.25, 0.5, 1, 2].map((w) => w * width),
  });

  return {
    height,
    src: normalizeSources(optimizedImage.src),
    srcSet: normalizeSources(optimizedImage.srcSet.attribute),
    width,
  };
}

export async function getOpenGraphCoverAsBase64String(work: Work) {
  const width = 420;
  const format = "png";
  let cacheKey = "";

  if (cacheConfig.enableCache) {
    await ensureCacheDir(cacheConfig.cacheDir);

    const cacheKeyData = `${work.slug}-${width}-${format}`;
    cacheKey = createCacheKey(cacheKeyData);

    const cachedCover = await getCachedItem<string>(
      cacheConfig.cacheDir,
      cacheKey,
      "txt",
      false,
      cacheConfig.debugCache,
      `Cover base64: ${work.slug}`,
    );

    if (cachedCover) {
      return cachedCover;
    }
  }

  const imageBuffer = await sharp(getWorkCoverPath(work))
    .resize(width)
    .toFormat(format)
    .toBuffer();

  const base64String = `data:${"image/png"};base64,${imageBuffer.toString("base64")}`;

  if (cacheConfig.enableCache) {
    await saveCachedItem(cacheConfig.cacheDir, cacheKey, "txt", base64String);
  }

  return base64String;
}

export async function getStructuredDataCoverSrc(work: Work): Promise<string> {
  const workCoverFile = await getWorkCoverFile(work);

  const optimizedImage = await getImage({
    format: "jpeg",
    height: 750,
    quality: 80,
    src: workCoverFile.default,
    width: 500,
  });

  return normalizeSources(optimizedImage.src);
}

export async function getUpdateCoverProps(
  work: Work,
): Promise<CoverImageProps> {
  const coverFile = await getWorkCoverFile(work);

  const coverFilePath = getWorkCoverPath(work);
  const height = await getCoverHeight(coverFilePath, 500);

  const optimizedImage = await getImage({
    format: "png",
    height,
    quality: 100,
    src: coverFile.default,
    width: 500,
  });

  return {
    height,
    src: normalizeSources(optimizedImage.src),
    srcSet: normalizeSources(optimizedImage.srcSet.attribute),
    width: 500,
  };
}

export function getWorkCoverPath(work: Work) {
  const workCover = coverPath(work.slug);

  if (workCover) {
    return workCover;
  }

  return coverPath("default") || "";
}

function coverPath(slug: string) {
  const coverPath = path.resolve(`./content/assets/covers/${slug}.png`);
  if (fs.existsSync(coverPath)) {
    return coverPath;
  }

  return;
}

async function getWorkCoverFile(work: Work) {
  const coverKey = Object.keys(images).find((image) => {
    return image.endsWith(`${work.slug}.png`);
  });

  if (coverKey) {
    return await images[coverKey]();
  }

  const defaultWorkCoverKey = Object.keys(images).find((image) => {
    return image.endsWith(`default.png`);
  })!;

  return await images[defaultWorkCoverKey]();
}
