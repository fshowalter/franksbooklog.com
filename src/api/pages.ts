import { remark } from "remark";
import remarkGfm from "remark-gfm";
import smartypants from "remark-smartypants";
import strip from "strip-markdown";

import { ENABLE_CACHE } from "~/utils/cache";
import { perfLogger } from "~/utils/performanceLogger";

import type { MarkdownPage as RawMarkdownPage } from "./data/pagesMarkdown";

import { allPagesMarkdown } from "./data/pagesMarkdown";
import {
  allReviewedWorksJson,
  type ReviewedWorkJson,
} from "./data/reviewedWorksJson";
import { getHtml } from "./utils/markdown/getHtml";
import { removeFootnotes } from "./utils/markdown/removeFootnotes";

let cachedPagesMarkdown: RawMarkdownPage[];
let cachedReviewedWorksJson: ReviewedWorkJson[];

type MarkdownPage = {
  content: string | undefined;
  rawContent: string;
  title: string;
};

export function getContentPlainText(rawContent: string): string {
  return getMastProcessor()
    .use(removeFootnotes)
    .use(strip)
    .processSync(rawContent)
    .toString();
}

export async function getPage(slug: string): Promise<MarkdownPage> {
  return await perfLogger.measure("getPage", async () => {
    const pages = cachedPagesMarkdown || (await allPagesMarkdown());
    if (ENABLE_CACHE && !cachedPagesMarkdown) {
      cachedPagesMarkdown = pages;
    }

    const matchingPage = pages.find((page) => {
      return page.slug === slug;
    })!;

    const reviewedWorksJson =
      cachedReviewedWorksJson || (await allReviewedWorksJson());
    if (ENABLE_CACHE && !cachedReviewedWorksJson) {
      cachedReviewedWorksJson = reviewedWorksJson;
    }

    return {
      content: getHtml(matchingPage.rawContent, reviewedWorksJson),
      rawContent: matchingPage?.rawContent || "",
      title: matchingPage.title,
    };
  });
}

function getMastProcessor() {
  return remark().use(remarkGfm).use(smartypants);
}
