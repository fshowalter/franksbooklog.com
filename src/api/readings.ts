import { perfLogger } from "~/utils/performanceLogger";

import type { ReadingEntryJson } from "./data/reading-entries-json";

import { allReadingEntriesJson } from "./data/reading-entries-json";

type ReadingEntries = {
  abandonedCount: number;
  bookCount: number;
  distinctEditions: string[];
  distinctKinds: string[];
  distinctReadingYears: string[];
  distinctWorkYears: string[];
  readingEntries: ReadingEntry[];
  shortStoryCount: number;
  workCount: number;
};

type ReadingEntry = ReadingEntryJson & {};

const yearFormatter = new Intl.DateTimeFormat("en-US", {
  timeZone: "UTC",
  year: "numeric",
});

export async function allReadingEntries(): Promise<ReadingEntries> {
  return await perfLogger.measure("allTimelineEntries", async () => {
    const readingEntries = await allReadingEntriesJson();
    const distinctWorkYears = new Set<string>();
    const distinctReadingYears = new Set<string>();
    const distinctKinds = new Set<string>();
    const distinctEditions = new Set<string>();
    const works = readingEntries.filter((entry) => {
      return entry.progress === "Finished" || entry.progress === "Abandoned";
    });

    for (const entry of readingEntries) {
      distinctEditions.add(entry.edition);
      distinctKinds.add(entry.kind);
      distinctReadingYears.add(
        yearFormatter.format(new Date(entry.readingEntryDate)),
      );
      distinctWorkYears.add(entry.workYear);
    }

    return {
      abandonedCount: works.filter((work) => work.progress === "Abandoned")
        .length,
      bookCount: works.filter((work) => work.kind !== "Short Story").length,
      distinctEditions: [...distinctEditions].toSorted(),
      distinctKinds: [...distinctKinds].toSorted(),
      distinctReadingYears: [...distinctReadingYears].toSorted(),
      distinctWorkYears: [...distinctWorkYears].toSorted(),
      readingEntries,
      shortStoryCount: works.filter((work) => work.kind === "Short Story")
        .length,
      workCount: works.length,
    };
  });
}
