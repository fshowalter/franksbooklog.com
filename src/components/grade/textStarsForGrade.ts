/**
 * Mapping of letter grades to their HTML entity star representations.
 * Used for generating text-based star ratings using Unicode star symbols
 * and fraction characters for half-star ratings.
 */
const gradeMap: Record<string, string> = {
  A: "&#9733;&#9733;&#9733;&#9733;&#9733;",
  "A+": "&#9733;&#9733;&#9733;&#9733;&#9733;",
  "A-": "&#9733;&#9733;&#9733;&#9733;&#189;",
  B: "&#9733;&#9733;&#9733;&#9733;",
  "B+": "&#9733;&#9733;&#9733;&#9733;",
  "B-": "&#9733;&#9733;&#9733;&#189;",
  C: "&#9733;&#9733;&#9733;",
  "C+": "&#9733;&#9733;&#9733;",
  "C-": "&#9733;&#9733;&#189;",
  D: "&#9733;&#9733;",
  "D+": "&#9733;&#9733;",
  "D-": "&#9733;&#189;",
  F: "&#9733;",
};

/**
 * Converts a letter grade to its HTML entity star representation.
 * Returns a string of HTML entities representing stars for the given grade.
 * 
 * @param grade - The letter grade (e.g., "A+", "B-", "C", etc.)
 * @returns HTML entity string representing the star rating
 */
export function textStarsForGrade(grade: string) {
  return gradeMap[grade];
}
