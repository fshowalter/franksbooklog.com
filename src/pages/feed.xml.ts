import rss from "@astrojs/rss";

import { getFeedCoverProps } from "~/api/covers";
import {
  loadExcerptHtml,
  mostRecentReviews,
  type ReviewWithExcerpt,
} from "~/api/reviews";
import { textStarsForGrade } from "~/components/grade/textStarsForGrade";

/**
 * Astro API endpoint that generates an RSS feed for the most recent book reviews.
 * Creates RSS items with cover images, grades, and excerpt content.
 *
 * @returns RSS response containing the latest 10 book reviews with metadata
 */
export async function GET() {
  const reviews = await mostRecentReviews(10);

  const rssItems = await Promise.all(
    reviews.map(async (review) => {
      return await loadExcerptHtml(review);
    }),
  );

  return rss({
    customData:
      "<image><url>https://www.franksbooklog.com/assets/favicon-128.png</url><title>Frank's Book Log</title><link>https://www.franksbooklog.com/</link></image>",
    // `<description>` field in output xml
    description: "Reviews of current, cult, classic, and forgotten books.",
    // Array of `<item>`s in output xml
    // See "Generating items" section for examples using content collections and glob imports
    items: await Promise.all(
      rssItems.map(async (item) => {
        const cover = await getFeedCoverProps(item);

        return {
          content: `<img src="${
            cover.src
          }" alt="">${addMetaToExcerpt(item.excerpt, item)}`,
          link: `https://www.franksbooklog.com/reviews/${item.slug}/`,
          pubDate: item.date,
          title: `${item.title} by ${authorsToString(item.authors)}`,
        };
      }),
    ),
    // Pull in your project "site" from the endpoint context
    // https://docs.astro.build/en/reference/api-reference/#contextsite
    site: "https://www.franksbooklog.com",
    // `<title>` field in output xml
    title: "Frank's Book Log",
  });
}

/**
 * Adds star rating metadata to the beginning of a review excerpt for RSS feed display.
 * Prepends the review's grade as text stars before the excerpt content.
 *
 * @param {string} excerpt - The HTML excerpt content from the review
 * @param {ReviewWithExcerpt} review - The review object containing grade and other metadata
 * @returns {string} HTML string with star rating metadata prepended to the excerpt
 */
function addMetaToExcerpt(excerpt: string, review: ReviewWithExcerpt) {
  const meta = `${textStarsForGrade(review.grade)}`;
  return `<p>${meta}</p>${excerpt}`;
}

/**
 * Converts an array of author objects to a human-readable string format.
 * Handles author names with optional notes and formats them using the browser's
 * Intl.ListFormat for proper comma separation and conjunction.
 *
 * @param {ReviewWithExcerpt["authors"]} authors - Array of author objects with name and optional notes
 * @returns {string} Formatted string of author names, e.g., "John Doe, Jane Smith (Editor), and Bob Johnson"
 */
function authorsToString(authors: ReviewWithExcerpt["authors"]) {
  const authorsArray = authors.map((author) => {
    if (author.notes) {
      return `${author.name} (${author.notes})`;
    }

    return author.name;
  });

  return new Intl.ListFormat().format(authorsArray);
}
