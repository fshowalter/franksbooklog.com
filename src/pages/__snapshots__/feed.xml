<?xml version="1.0" encoding="UTF-8" ?>
<rss
  version="2.0"
  xmlns:content="http://purl.org/rss/1.0/modules/content/"
><channel><title>Frank&apos;s Book Log</title><description
    >Reviews of current, cult, classic, and forgotten books.</description><link
    >https://www.franksbooklog.com/</link><image><url
      >https://www.franksbooklog.com/assets/favicon-128.png</url><title
      >Frank&apos;s Book Log</title><link
      >https://www.franksbooklog.com/</link></image><item><title
      >Ghost Story by Peter Straub</title><link
      >https://www.franksbooklog.com/reviews/ghost-story-by-peter-straub/</link><guid
        isPermaLink="true"
      >https://www.franksbooklog.com/reviews/ghost-story-by-peter-straub/</guid><pubDate
      >Sun, 01 Dec 2024 00:00:00 GMT</pubDate><content:encoded
      >&lt;img src=&quot;/_image/?href=%2Fcontent%2Fassets%2Fcovers%2Fghost-story-by-peter-straub.png%3ForigWidth%3D500%26origHeight%3D750%26origFormat%3Dpng&amp;w=500&amp;h=750&amp;q=80&amp;f=jpeg&quot; alt=&quot;&quot;&gt;&lt;p&gt;&amp;#9733;&amp;#9733;&amp;#9733;&amp;#9733;&lt;/p&gt;&lt;p&gt;&lt;em&gt;Ghost Story&lt;/em&gt; is an uneven—yet at times startlingly effective—supernatural horror with memorable characters and a vivid small-town setting. Just don’t expect traditional ghosts. &lt;a href=&quot;/reviews/ghost-story-by-peter-straub/&quot;&gt;Read more...&lt;/a&gt;&lt;/p&gt;</content:encoded></item><item
    ><title>Goth by Lol Tolhurst</title><link
      >https://www.franksbooklog.com/reviews/goth-by-lol-tolhurst/</link><guid
        isPermaLink="true"
      >https://www.franksbooklog.com/reviews/goth-by-lol-tolhurst/</guid><pubDate
      >Fri, 08 Nov 2024 00:00:00 GMT</pubDate><content:encoded
      >&lt;img src=&quot;/_image/?href=%2Fcontent%2Fassets%2Fcovers%2Fgoth-by-lol-tolhurst.png%3ForigWidth%3D500%26origHeight%3D750%26origFormat%3Dpng&amp;w=500&amp;h=750&amp;q=80&amp;f=jpeg&quot; alt=&quot;&quot;&gt;&lt;p&gt;&amp;#9733;&amp;#9733;&amp;#9733;&lt;/p&gt;&lt;p&gt;Former Cure drummer and keyboardist Lol Tolhurst chronicles the Goth movement from its historical roots, through its post-punk inception to modern day. &lt;a href=&quot;/reviews/goth-by-lol-tolhurst/&quot;&gt;Read more...&lt;/a&gt;&lt;/p&gt;</content:encoded></item><item
    ><title>Recessional by David Mamet</title><link
      >https://www.franksbooklog.com/reviews/recessional-by-david-mamet/</link><guid
        isPermaLink="true"
      >https://www.franksbooklog.com/reviews/recessional-by-david-mamet/</guid><pubDate
      >Sun, 25 Aug 2024 00:00:00 GMT</pubDate><content:encoded
      >&lt;img src=&quot;/_image/?href=%2Fcontent%2Fassets%2Fcovers%2Frecessional-by-david-mamet.png%3ForigWidth%3D500%26origHeight%3D750%26origFormat%3Dpng&amp;w=500&amp;h=750&amp;q=80&amp;f=jpeg&quot; alt=&quot;&quot;&gt;&lt;p&gt;&amp;#9733;&amp;#9733;&lt;/p&gt;&lt;p&gt;Non-politicians like David Mamet, who’ve done a hard pivot from one political extreme to the other, fascinate me. In an effort to understand his mindset, I picked up &lt;em&gt;Recessional&lt;/em&gt;. &lt;a href=&quot;/reviews/recessional-by-david-mamet/&quot;&gt;Read more...&lt;/a&gt;&lt;/p&gt;</content:encoded></item><item
    ><title>Fast Forward by Stephen Morris</title><link
      >https://www.franksbooklog.com/reviews/fast-forward-by-stephen-morris/</link><guid
        isPermaLink="true"
      >https://www.franksbooklog.com/reviews/fast-forward-by-stephen-morris/</guid><pubDate
      >Mon, 01 Jul 2024 00:00:00 GMT</pubDate><content:encoded
      >&lt;img src=&quot;/_image/?href=%2Fcontent%2Fassets%2Fcovers%2Ffast-forward-by-stephen-morris.png%3ForigWidth%3D500%26origHeight%3D750%26origFormat%3Dpng&amp;w=500&amp;h=750&amp;q=80&amp;f=jpeg&quot; alt=&quot;&quot;&gt;&lt;p&gt;&amp;#9733;&amp;#9733;&amp;#9733;&amp;#9733;&amp;#189;&lt;/p&gt;&lt;p&gt;Every bit as entertaining as his first volume, drummer Stephen Morris’s second volume of memoirs picks up where &lt;em&gt;Record Play Pause&lt;/em&gt; ended, chronicling New Order’s rise to super-stardom and the tensions that tore it apart. &lt;a href=&quot;/reviews/fast-forward-by-stephen-morris/&quot;&gt;Read more...&lt;/a&gt;&lt;/p&gt;</content:encoded></item><item
    ><title>Gothic by Philip Fracassi</title><link
      >https://www.franksbooklog.com/reviews/gothic-by-philip-fracassi/</link><guid
        isPermaLink="true"
      >https://www.franksbooklog.com/reviews/gothic-by-philip-fracassi/</guid><pubDate
      >Sun, 30 Jun 2024 00:00:00 GMT</pubDate><content:encoded
      >&lt;img src=&quot;/_image/?href=%2Fcontent%2Fassets%2Fcovers%2Fgothic-by-philip-fracassi.png%3ForigWidth%3D500%26origHeight%3D750%26origFormat%3Dpng&amp;w=500&amp;h=750&amp;q=80&amp;f=jpeg&quot; alt=&quot;&quot;&gt;&lt;p&gt;&amp;#9733;&amp;#9733;&amp;#9733;&amp;#9733;&lt;/p&gt;&lt;blockquote&gt;
&lt;p&gt;“It’s like Christine… but wood!”&lt;/p&gt;
&lt;/blockquote&gt;
&lt;p&gt;So remarks a character in Philip Fracassi’s meta horror novel about a haunted desk. And while the statement is true per se, it doesn’t do justice to Fracassi’s work, which proves far leaner and meaner than King’s. &lt;a href=&quot;/reviews/gothic-by-philip-fracassi/&quot;&gt;Read more...&lt;/a&gt;&lt;/p&gt;</content:encoded></item><item
    ><title>Record Play Pause by Stephen Morris</title><link
      >https://www.franksbooklog.com/reviews/record-play-pause-by-stephen-morris/</link><guid
        isPermaLink="true"
      >https://www.franksbooklog.com/reviews/record-play-pause-by-stephen-morris/</guid><pubDate
      >Sat, 11 Nov 2023 00:00:00 GMT</pubDate><content:encoded
      >&lt;img src=&quot;/_image/?href=%2Fcontent%2Fassets%2Fcovers%2Frecord-play-pause-by-stephen-morris.png%3ForigWidth%3D500%26origHeight%3D750%26origFormat%3Dpng&amp;w=500&amp;h=750&amp;q=80&amp;f=jpeg&quot; alt=&quot;&quot;&gt;&lt;p&gt;&amp;#9733;&amp;#9733;&amp;#9733;&amp;#9733;&amp;#9733;&lt;/p&gt;&lt;p&gt;Warm, witty, self-deprecating and self-aware, drummer Stephen Morris’s first volume of memoirs chronicles his formative years in the Manchester suburb of Macclesfield through joining Joy Division, its end, and the early days of New Order. &lt;a href=&quot;/reviews/record-play-pause-by-stephen-morris/&quot;&gt;Read more...&lt;/a&gt;&lt;/p&gt;</content:encoded></item><item
    ><title>Fright Night Origins by Tom Holland and A. Jack Ulrich</title><link
      >https://www.franksbooklog.com/reviews/fright-night-origins-by-tom-holland-a-jack-ulrich/</link><guid
        isPermaLink="true"
      >https://www.franksbooklog.com/reviews/fright-night-origins-by-tom-holland-a-jack-ulrich/</guid><pubDate
      >Mon, 06 Nov 2023 00:00:00 GMT</pubDate><content:encoded
      >&lt;img src=&quot;/_image/?href=%2Fcontent%2Fassets%2Fcovers%2Ffright-night-origins-by-tom-holland-a-jack-ulrich.png%3ForigWidth%3D500%26origHeight%3D750%26origFormat%3Dpng&amp;w=500&amp;h=750&amp;q=80&amp;f=jpeg&quot; alt=&quot;&quot;&gt;&lt;p&gt;&amp;#9733;&amp;#9733;&amp;#9733;&lt;/p&gt;&lt;p&gt;Tom Holland, director and screenwriter of the original &lt;em&gt;Fright Night&lt;/em&gt; film, partners with A. Jack Ulrich to craft an expanded novelization of the film, proffering more backstory and setting up an expanded universe and sequel. &lt;a href=&quot;/reviews/fright-night-origins-by-tom-holland-a-jack-ulrich/&quot;&gt;Read more...&lt;/a&gt;&lt;/p&gt;</content:encoded></item><item
    ><title>Don&apos;t Let Them Get You Down by Philip Fracassi</title><link
      >https://www.franksbooklog.com/reviews/dont-let-them-get-you-down-by-philip-fracassi/</link><guid
        isPermaLink="true"
      >https://www.franksbooklog.com/reviews/dont-let-them-get-you-down-by-philip-fracassi/</guid><pubDate
      >Sun, 29 Oct 2023 00:00:00 GMT</pubDate><content:encoded
      >&lt;img src=&quot;/_image/?href=%2Fcontent%2Fassets%2Fcovers%2Fdont-let-them-get-you-down-by-philip-fracassi.png%3ForigWidth%3D500%26origHeight%3D750%26origFormat%3Dpng&amp;w=500&amp;h=750&amp;q=80&amp;f=jpeg&quot; alt=&quot;&quot;&gt;&lt;p&gt;&amp;#9733;&amp;#9733;&amp;#189;&lt;/p&gt;&lt;p&gt;&lt;em&gt;Don’t Let Them Get You Down&lt;/em&gt; follows Peter, a copywriter for a medical website, struggling to hold his life together while battling a spiraling depression triggered by his wife’s cancer death two years prior. &lt;a href=&quot;/reviews/dont-let-them-get-you-down-by-philip-fracassi/&quot;&gt;Read more...&lt;/a&gt;&lt;/p&gt;</content:encoded></item><item
    ><title>The Creative Act by Rick Rubin</title><link
      >https://www.franksbooklog.com/reviews/the-creative-act-by-rick-rubin/</link><guid
        isPermaLink="true"
      >https://www.franksbooklog.com/reviews/the-creative-act-by-rick-rubin/</guid><pubDate
      >Sun, 22 Oct 2023 00:00:00 GMT</pubDate><content:encoded
      >&lt;img src=&quot;/_image/?href=%2Fcontent%2Fassets%2Fcovers%2Fthe-creative-act-by-rick-rubin.png%3ForigWidth%3D500%26origHeight%3D750%26origFormat%3Dpng&amp;w=500&amp;h=750&amp;q=80&amp;f=jpeg&quot; alt=&quot;&quot;&gt;&lt;p&gt;&amp;#9733;&amp;#9733;&amp;#9733;&lt;/p&gt;&lt;p&gt;Rick Rubin, the famed music producer of such disparate acts as the Beastie Boys, Slayer, Tom Petty, and Johnny Cash, proffers a meditation on creativity. &lt;a href=&quot;/reviews/the-creative-act-by-rick-rubin/&quot;&gt;Read more...&lt;/a&gt;&lt;/p&gt;</content:encoded></item><item
    ><title>Sacculina by Philip Fracassi</title><link
      >https://www.franksbooklog.com/reviews/sacculina-by-philip-fracassi/</link><guid
        isPermaLink="true"
      >https://www.franksbooklog.com/reviews/sacculina-by-philip-fracassi/</guid><pubDate
      >Mon, 09 Oct 2023 00:00:00 GMT</pubDate><content:encoded
      >&lt;img src=&quot;/_image/?href=%2Fcontent%2Fassets%2Fcovers%2Fsacculina-by-philip-fracassi.png%3ForigWidth%3D500%26origHeight%3D750%26origFormat%3Dpng&amp;w=500&amp;h=750&amp;q=80&amp;f=jpeg&quot; alt=&quot;&quot;&gt;&lt;p&gt;&amp;#9733;&amp;#9733;&amp;#9733;&amp;#9733;&lt;/p&gt;&lt;p&gt;To celebrate Jim’s older brother Jack’s release following a six-year prison stint, the brothers, their father Henry, and Jack’s hulking best friend Chris, charter a fishing boat in the Pacific. The choppy seas don’t bode well for an outing, but the group presses ahead, and their grizzled captain takes them a few miles offshore to a deep spot good for fishing. But the men soon realize they’re under threat by an unknown, overwhelming assailant—and how alone they are at sea. &lt;a href=&quot;/reviews/sacculina-by-philip-fracassi/&quot;&gt;Read more...&lt;/a&gt;&lt;/p&gt;</content:encoded></item></channel></rss>
